awx/ui/src/screens/Job/JobOutput/useJobEvents.js

```diff
@@ -409,10 +423,37 @@ function getNumCollapsedChildren(node) {
 }
 
 function toggleNodeIsCollapsed(state, eventUuid) {
-  return updateNodeByUuid(state, eventUuid, (node) => ({
+  return {
+    ...updateNodeByUuid(state, eventUuid, (node) => ({
+      ...node,
+      isCollapsed: !node.isCollapsed,
+    })),
+    isAllCollapsed: state.isAllCollapsed ? false : state.isAllCollapsed,
+  };
+}
+
+function toggleCollapseAll(state, isAllCollapsed) {
+  const newTree = state.tree.map((node) =>
+    _toggleNestedNodes(state.events, node, isAllCollapsed)
+  );
+  return { ...state, tree: newTree, isAllCollapsed };
+}
+
+function _toggleNestedNodes(events, node, isCollapsed) {
+  const eventShouldNotCollapse =
+    events[node.eventIndex].event_data.playbook_uuid ===
+      events[node.eventIndex].event_data.parent_uuid ||
+    !events[node.eventIndex].parent_uuid.length;
```

Comment:
minor personal preference thing: I would set `const event = events[node.eventIndex]` before this and use that var here to make the logic a bit cleaner to read